service/qos/service_level_controller.cc

```diff
@@ -87,14 +88,21 @@ void service_level_controller::set_distributed_data_accessor(service_level_distr
 }
 
 future<> service_level_controller::stop() {
-    // unregister from the service level distributed data accessor.
-    _sl_data_accessor = nullptr;
-    if (this_shard_id() == global_controller) {
-        // abort the loop of the distributed data checking if it is running
-        _global_controller_db->dist_data_update_aborter.request_abort();
-        _global_controller_db->notifications_serializer.broken();
+    try {
+        // unregister from the service level distributed data accessor.
+        _sl_data_accessor = nullptr;
+        if (this_shard_id() == global_controller) {
+            // abort the loop of the distributed data checking if it is running
+            _global_controller_db->dist_data_update_aborter.request_abort();
+            _global_controller_db->notifications_serializer.broken();
+            co_await std::exchange(_global_controller_db->distributed_data_update, make_ready_future<>());
+        }
+        co_await std::exchange(_distributed_data_updater, make_ready_future<>());
+    } catch (const broken_semaphore&) {
+        // semaphore errors are expected
```

Comment:
A question:
If the first co_await throws, the second co_await is not done at all. Is this fine or impossible (i.e. the first one doesn't throw)?
If it isn't fine, you need to use one of the idioms that waits for multiple futures, and probably not coroutines :-(

Reply:
If the first co_await throws, then the second one is already ready in this specific situation, so it's safe to discard. And if the second one failed, the first one would never be assigned to, so it will be a ready future and it will not throw.

Reply:
Ok, I'll take your word for it - I don't understand when are these broken semaphore et al. exceptions are expected, and why it's enough to only wait for one of these futures in that case, but you obviously know the flow and are correct.

Reply:
For the sake of being future-proof, I tend to agree with @nyh that we better `co_await when_all_succeed()`
on both, if `_distributed_data_updater` should be ready in case `_global_controller_db->distributed_data_update` failed.

Reply:
sure, I'll rework and send v2